pkg/cloudprovider/providers/azure/azure_test.go

```diff
@@ -1045,9 +1564,29 @@ func TestMetadataParsing(t *testing.T) {
 	}
 }
 
-func addTestSubnet(t *testing.T, svc *v1.Service) {
+func addTestSubnet(t *testing.T, az *Cloud, svc *v1.Service) {
 	if svc.Annotations[ServiceAnnotationLoadBalancerInternal] != "true" {
 		t.Error("Subnet added to non-internal service")
 	}
-	svc.Annotations[ServiceAnnotationLoadBalancerInternalSubnet] = "TestSubnet"
+	subName := svc.Annotations[ServiceAnnotationLoadBalancerInternalSubnet]
+	if subName == "" {
+		subName = az.SubnetName
+	}
+
+	subnetID := fmt.Sprintf("/subscriptions/%s/resourceGroups/%s/providers/Microsoft.Network/virtualNetworks/%s/subnets/%s",
+		az.SubscriptionID,
+		az.VnetResourceGroup,
+		az.VnetName,
+		subName)
+
+	_, errChan := az.SubnetsClient.CreateOrUpdate(az.VnetResourceGroup, az.VnetName, subName,
```

Comment:
this feels like we are forced to add complexity to deal with the mixing of the pure and impure code - I would be keen to revert as far as possible to the 'pure calculation + impure application' separation

Reply:
I think that mostly about a personal preference. 
I don't see why we need to separate the calculation and application. 

And I also, i did cross reference on gce and aws's implementation. 
Specifically for EnsureLoadBalancer function, they bind calculation and application into one function for sure. Feel free to take a look.

So we won't do revert for sure.